stl/inc/algorithm

```diff
@@ -3851,10 +3851,17 @@ _CONSTEXPR20 _OutIt replace_copy(_InIt _First, _InIt _Last, _OutIt _Dest, const
     const auto _ULast = _STD _Get_unwrapped(_Last);
     auto _UDest       = _STD _Get_unwrapped_n(_Dest, _STD _Idl_distance<_InIt>(_UFirst, _ULast));
     for (; _UFirst != _ULast; ++_UFirst, (void) ++_UDest) {
-        if (*_UFirst == _Oldval) {
-            *_UDest = _Newval;
+        if constexpr ((is_integral_v<_Ty> && is_integral_v<_Iter_value_t<_InIt>>)
+                      || (is_pointer_v<_Ty> && is_pointer_v<_Iter_value_t<_InIt>>) ) {
+            // TRANSITION, DevCom-10606350: help the compiler auto-vectorizing for simple types
+            // by making transformation that can be done for simple types
+            *_UDest = (*_UFirst == _Oldval) ? _Newval : *_UFirst;

```

Comment:
I don't believe that this is an allowable transformation. WG21-N4971 \[alg.replace\]/7 "*Mandates:* The results of the expressions `*first` and `new_value` are writable (\[iterator.requirements.general\]) to `result`." If `_Ty` is `unsigned char` and `*_UFirst` is `unsigned short`, then the conditional operator will perform the usual arithmetic conversions and we'll try to write `int` to the output iterator, which might not compile.

For pointers, we'll get their composite pointer type, which can similarly synthesize a third type.

Similarly for the ranges algorithm below, we have `output_iterator<_Out, const _Ty2&>` and `indirectly_copyable<_It, _Out>` but not whatever the usual arithmetic conversions might try to do.

We could add even more metaprogramming to detect output iterators which won't care about the difference, but I think that's too much library complexity to attempt to compensate for an optimizer deficiency. Thanks for investigating this (and reporting the optimizer issue!), but I don't think we should make any changes here.

Reply:
We already do some complex metaprogramming elsewhere.
Consider `_Equal_memcmp_is_safe`, `_Lex_compare_three_way_memcmp_classify` and `_Vector_alg_in_find_is_safe`.
In those cases we direct optimized path to `memcmp` function or our own vector algorithm, so ultimately doing more complex thing. I think directing to auto-vectorization is more favorable.

Reply:
I'm not super eager to take on additional maintenance burden and regression risk to compensate for C2's current deficiency (as I observe that LLVM happily autovectorizes this), but I think I'd be okay with allowing a limited set of easy-to-reason-about scenarios through. I think we need (in addition to the current criteria):

* The sources are the same type (so the conditional operator won't synthesize a third type)
* The unwrapped output iterator is a raw pointer to an integer/pointer (practically always necessary for autovectorization anyways, and this ensures that the output iterator won't be playing squirrelly games with inspecting constness/value category)

Given those restrictions, I don't think we need to require the output type to be the same as the source types.

Reply:
One more thing to consider here is #4433 . If it is resolved to be desirable optimization, we would basically share the metaprogramming.

Otherwise I'd want the compiler to fix deficiency, but whereas such issues ocassionally fixed, and sometimes even quickly, most of such issues created by me are either still open, or closed without fix.

Reply:
We talked about this at the weekly maintainer meeting and @barcharcraz says: "This is gross, but grr, fine" :joy_cat: :zany_face: